Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1556 @@
+  erw [max_eq_eval_unapply C hsC ho]
+  intro h
+  let f := MaxToGood C hC hsC ho h₁
+  have hf : f.Injective := maxToGood_injective C hC hsC ho h₁
+  have hh : (fun l ↦ Products.eval (C' C ho) l.val.Tail) = eval (C' C ho) ∘ f := rfl
+  rw [hh]
+  exact h.comp f hf
+
+end GoodProducts
+
+end GoodProducts
+
+end Successor
+
+section Induction
+
+variable (I) in
+/-- The predicate on ordinals which we prove by induction, see `GoodProducts.P0`,
+    `GoodProducts.Plimit` and `GoodProducts.linearIndependentAux`.  -/
+def P (o : Ordinal) : Prop :=
+  o ≤ Ordinal.type (·<· : I → I → Prop) →
+  (∀ (C : Set (I → Bool)), IsClosed C → contained C o →
+    LinearIndependent ℤ (GoodProducts.eval C))
+
+theorem GoodProducts.P0 : P I 0 := fun _ C _ hsC ↦ by
+  have : C ⊆ {(fun _ ↦ false)} := fun c hc ↦ by
+    ext x; exact Bool.eq_false_iff.mpr (fun ht ↦ (Ordinal.not_lt_zero (ord I x)) (hsC c hc x ht))
+  rw [Set.subset_singleton_iff_eq] at this
+  cases this
+  · subst C
+    exact linearIndependentEmpty
+  · subst C
+    exact linearIndependentSingleton
+
+theorem GoodProducts.Plimit (o : Ordinal) (ho : Ordinal.IsLimit o) :
+    (∀ (o' : Ordinal), o' < o → P I o') → P I o := by
+  intro h hho C hC hsC
+  rw [linearIndependent_iff_union_smaller C ho hsC]
+  exact linearIndependent_iUnion_of_directed
+    (Monotone.directed_le fun _ _ h ↦ GoodProducts.smaller_mono C h) fun ⟨o', ho'⟩ ↦
+    (linearIndependent_iff_smaller _ _).mp (h o' ho' (le_of_lt (lt_of_lt_of_le ho' hho))
+    (π C (ord I · < o')) (isClosed_proj _ _ hC) (contained_proj _ _))
+
+theorem GoodProducts.linearIndependentAux (μ : Ordinal) : P I μ := by
+  refine Ordinal.limitRecOn μ P0 (fun o h ho C hC hsC ↦ ?_)
+      (fun o ho h ↦ (GoodProducts.Plimit o ho (fun o' ho' ↦ (h o' ho'))))
+  have ho' : o < Ordinal.type (·<· : I → I → Prop) :=
+    lt_of_lt_of_le (Order.lt_succ _) ho
+  rw [linearIndependent_iff_sum C hsC ho']
+  refine ModuleCat.linearIndependent_leftExact ?_ ?_ (succ_mono C o) (succ_exact C hC hsC ho')
+      (square_commutes C ho')
+  · exact h (le_of_lt ho') (π C (ord I · < o)) (isClosed_proj C o hC) (contained_proj C o)
+  · exact linearIndependent_comp_of_eval C hC hsC ho' (span (π C (ord I · < o))
+      (isClosed_proj C o hC)) (h (le_of_lt ho') (C' C ho') (isClosed_C' C hC ho')
+      (contained_C' C ho'))
+
+theorem GoodProducts.linearIndependent (hC : IsClosed C) :
+    LinearIndependent ℤ (GoodProducts.eval C) :=
+  GoodProducts.linearIndependentAux (Ordinal.type (·<· : I → I → Prop)) (le_refl _)
+    C hC (fun _ _ _ _ ↦ Ordinal.typein_lt_type _ _)
+
+/-- `GoodProducts C` as a `ℤ`-basis for `LocallyConstant C ℤ`.  -/
+noncomputable
+def GoodProducts.Basis (hC : IsClosed C) :
+    Basis (GoodProducts C) ℤ (LocallyConstant C ℤ) :=
+  Basis.mk (GoodProducts.linearIndependent C hC) (GoodProducts.span C hC)
+
+end Induction
+
+variable {S : Profinite} {ι : S → I → Bool} (hι : ClosedEmbedding ι)
+
+theorem Nobeling : Module.Free ℤ (LocallyConstant S ℤ) := Module.Free.of_equiv'
```

Comment:
You call it Nobeling and again no docstring (so people browsing the code will be confused). Big docstring stating the theorem. Also: the type of Nobeling involves iota and hiota, this should perhaps be called Nobeling in your namespace. You could add a docstring explaining why it's not the final version and directing people to the stronger claim.